### ISSUE TITLE
Solang does not implement address.code

### ISSUE THREAD

**Issue by LucasSte** [hyperledger-labs/solang#809]
This contract works on Solc, but fails on Solang

```solidity
contract Code {

    function isContract(address account) internal view returns (bool) {
        // This method relies on extcodesize/address.code.length, which returns 0
        // for contracts in construction, since the code is only stored at the end
        // of the constructor execution.

        return account.code.length > 0;
    }

}
```

The error is the following:
```
error: 'code' not found
  ┌─ /home/lucas/Documents/solang/examples/array.sol:8:16
  │
8 │         return account.code.length > 0;
  │                ^^^^^^^^^^^^

```

**Comment by oblx** [comment]
Same here while compiling an ERC721.

Did you find a workaround @LucasSte ?

**Comment by LucasSte** [comment]
> 

We haven't found a solution yet. Can you provide more details on what you're trying to achieve? We are certainly can find a way around.

If you are interesting in creating NFTs for Solana, for instance, you need to use the SPL-token integration and not ERC721. You can find more information about how to use the SPL-token on Solana and Solang [here](https://solang.readthedocs.io/en/latest/targets/solana.html#using-spl-token).

**Comment by oblx** [comment]
> 

Hi and thanks for your answer !
I was dumbly trying to convert an erc721 contract and switch to solana like I was always there. But no, definitely no.

Thanks for the doc, I definitely need more knowledge before deploying on solana :)

**Comment by xermicus** [comment]
I'm going to pick this up, [we need it](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/cb4bf950df5ae43356c4935b3900446f6dc20261/contracts/proxy/ERC1967/ERC1967Utils.sol#L67).

**Comment by bkushigian** [comment]
@xermicus thanks for linking to this issue! I didn't see it (probably should have dug through the issue tracker before spamming you guys! Sorry, I was in go mode :D)

So I think #1537 is slightly different than this one, though they're obviously related. This issue is that solang's pipeline fails on `address.code` while my issue is about the parse and resolve component of the pipeline. I think that even if solang can't handle `address.code`, the AST should still be built and errors should be pushed down the pipeline.

**Comment by bkushigian** [comment]
I'm coming at this as a client of only `parse_and_resolve`, I just need an AST (with types, so parsetree is not enough) and I never touch anything else in the pipeline.

**Comment by bkushigian** [comment]
@xermicus just want to ping you on this. Would it be fair to reopen #1537 since it's a subproblem that could be fixed even if this issue is unfixed? The sema stage should not fail because of an `address.code`, even if the rest of the pipeline cannot handle it.

**Comment by xermicus** [comment]
@bkushigian yeah good point, it certainly can be implemented in sema but not in later stages. But let's just keep this issue only, implementing it in sema is part of it anyways :upside_down_face:

**Comment by bkushigian** [comment]
@xermicus I guess my worry is that this issue has been open for over a year, and that indicates it's probably kinda difficult to fix. Fixing the sema phase should be relatively straightforward to fix (compared to fixing the full pipeline), but if it doesn't have its own issue I don't see this being fixed until the entire bug gets fixed.

Context for why I care: I am trying to record the bugs in sema that prevent an AST from building. To do this I'm running sema on a bunch of real-world code to determine how often sema fails (currently >50%, as I [report here](https://discord.com/channels/905194001349627914/905834552965103638/1153351987417587913)). The `address.code` issue is pretty prevalent (I don't have hard numbers, but I've seen it occur quite a bit).

**Comment by xermicus** [comment]
@bkushigian FIY I implemented it for sema in #1549

**Comment by bkushigian** [comment]
@xermicus thank you!

**Comment by xermicus** [comment]
Removing from the OZ milestone for Polkadot; OppenZeppelin contracts check the `address.code.length` to distinguish contract addresses, however on Polkadot we have the `is_contract` builtin to do the same.
